packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -238,43 +308,150 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
             fundingRateRecord.proposeTime = proposalTime;
-            emit PublishedRate(perpetual, settlementRate.rawValue, proposalTime, proposer);
+            // Note: Set rewards to 0 since we published the DVM resolved funding rate and there is no proposer to pay.
+            emit PublishedRate(perpetual, settlementRate.rawValue, proposalTime, proposer, 0, 0, 0);
         }
 
         // Delete dispute
         delete fundingRateDisputes[perpetual][proposalTime];
     }
 
+    /**
+     * @notice Publishes any expired pending proposals and pays out proposal rewards if neccessary.
+     * @param perpetual Contract to check proposed funding rates for.
+     */
+    function withdrawProposalRewards(address perpetual) external nonReentrant() publishAndWithdrawProposal(perpetual) {}
+
+    /**
+     * @notice Set the reward rate (per second) for a specific `perpetual` contract.
+     * @dev Callable only by the Perpetual contract.
+     */
+    function setRewardRate(address perpetual, FixedPoint.Unsigned memory rewardRate)
+        external
+        override
+        nonReentrant()
+        publishAndWithdrawProposal(perpetual)
+    {
+        require(msg.sender == perpetual, "Caller not perpetual");
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
+        fundingRateRecord.rewardRatePerSecond = rewardRate;
+
+        // Set last propose time to current time since rewards will start accruing from here on out.
+        fundingRateRecord.proposeTime = getCurrentTime();
+
+        emit ChangedRewardRate(perpetual, rewardRate.rawValue);
+    }
+
+    /**
+     * @notice Helpful method for proposers who want to calculate their potential rewards and useful for testing.
+     */
+    function calculateProposalRewardPct(
+        address perpetual,
+        uint256 startTime,
+        uint256 endTime,
+        FixedPoint.Signed memory proposedRate,
+        FixedPoint.Signed memory currentRate
+    ) external view returns (FixedPoint.Unsigned memory reward) {
+        return _calculateProposalRewardPct(perpetual, startTime, endTime, proposedRate, currentRate);
+    }
+
     /****************************************
      *         INTERNAL FUNCTIONS           *
      ****************************************/
 
+    function _publishRateAndWithdrawRewards(address perpetual) internal {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
+
+        // Check if proposal liveness has expired
+        if (_getProposalState(fundingRateRecord.proposal) != ProposalState.Expired) {
+            return;
+        }
+
+        // Publish rate and proposal time.
+        fundingRateRecord.rate = fundingRateRecord.proposal.rate;
+        fundingRateRecord.proposeTime = fundingRateRecord.proposal.time;
+
+        // Calculate reward for proposer using saved reward rate and current perpetual PfC.
+        FixedPoint.Unsigned memory pfc = AdministrateeInterface(perpetual).pfc();
```

Comment:
The use of pfc in this equation like this makes this implementation and every function in the funding rate store associated with rewards vulnerable to a flash loan attack! Specifically, because the reward is a function of pfc, a flash swap can be used to increase the perp pfc during the withdrawal tx and to extract an unfair reward for proposing a funding rate. 

This attack would look as follows:
1) Propose a valid funding rate.
2) wait until liveness period has passed.
3) execute a transaction that does the following internal steps:
3.a) flash swap perp collateral currency.
3.b) create a small position, well above the GCR, collateralized with the flashswap'ed collateral. The larger the relative size of the flashswap to the total collateral in the contract the more profitable the attack.
3.c) withdraw your reward from the proposal. During this, the calculation of pfc is biased upwards from your collateral created during 3.b.
3.d) redeem position created in 3.b. This will return all flashborrowed collateral
3.e) repay flashloan
4) profit

In a discussion with @nicholaspai there is a solution to this by enforcing that the last addition to contract pfc was done in a separate block to `_publishRateAndWithdrawRewards` being called. This solution is discussed in issue https://github.com/UMAprotocol/protocol/issues/2183 in more detail. While this is a breaking issue that **must** be addressed before this contract is audited, it should be left for a separate PR due to the implementation complexity in this proposal.

Reply:
Great point -- I hadn't considered these sorts of attacks. +1. However, I don't _think_ this is a vulnerability (I think there's a minor flaw in your analysis). Will explain in the issue.

Reply:
As discussed IRL this is not a vulnerability upon further thought as the attacker would not gain marginal net profit